File: web/src/data/adminRequests.ts
/**
 * A4 — the support browser.
 *
 * Its own module rather than another thousand lines in `admin.ts`: nothing
 * here is shared with the accounts screen except the audit key it invalidates.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { AUDIT_KEY, STATS_KEY } from '@/data/admin'
import { api } from '@/data/client'
import type { JobStatus } from '@/data/jobs'
import type { RequestStatus, Urgency } from '@/data/requests'
import type { City, Page, Trade } from '@/data/types'

export const REQUESTS_PER_PAGE = 25

export const ADMIN_REQUESTS_KEY = ['admin', 'requests'] as const

export interface RequestPerson {
  id: number
  full_name: string
  /** Here because A4 exists while somebody is *on* the phone. An admin who has
   *  to leave the screen to find the number reads it out wrong. */
  phone: string
}

export interface AdminRequestRow {
  id: number
  title: string
  status: RequestStatus
  urgency: Urgency
  offers_count: number
  created_at: string
  client: RequestPerson
  trade: Trade
  city: City
}

export interface AdminRequestOffer {
  id: number
  price_centimes: number
  message: string | null
  status: 'pending' | 'accepted' | 'rejected' | 'withdrawn' | 'expired'
  created_at: string
  /** Frozen when it was accepted; null on an offer nobody took. */
  lead_fee_centimes: number | null
  provider: RequestPerson
  provider_id: number
}

export interface AdminRequestJob {
  id: number
  status: JobStatus
  agreed_price_centimes: number
  started_at: string | null
  finished_at: string | null
  confirmed_at: string | null
  cancelled_at: string | null
  cancelled_by: string | null
  cancel_reason: string | null
  provider: RequestPerson | null
}

export interface AdminRequestDispute {
  id: number
  status: 'open' | 'claimed' | 'resolved'
  reason: string
  opened_by_id: number
}

export interface AdminRequest {
  id: number
  title: string
  description: string
  status: RequestStatus
  urgency: Urgency
  address: string
  budget_min_centimes: number | null
  budget_max_centimes: number | null
  photos: number
  offers_count: number

  created_at: string
  expires_at: string | null
  cancelled_at: string | null
  cancel_reason: string | null

  client: RequestPerson
  trade: Trade
  city: City

  offers: AdminRequestOffer[]
  job: AdminRequestJob | null
  dispute: AdminRequestDispute | null

  /** Answered by the API rather than derived here from `status`, so the button
   *  and the endpoint agree about what "open" means. */
  can_cancel: boolean
}

export interface RequestFilters {
  q: string
  status: RequestStatus | ''
  cityId: number | ''
  tradeId: number | ''
}

export function useAdminRequests(filters: RequestFilters, page: number) {
  const params = new URLSearchParams({
    per_page: String(REQUESTS_PER_PAGE),
    page: String(page),
  })
  if (filters.q.trim()) params.set('q', filters.q.trim())
  if (filters.status) params.set('status', filters.status)
  if (filters.cityId) params.set('city_id', String(filters.cityId))
  if (filters.tradeId) params.set('trade_id', String(filters.tradeId))

  return useQuery({
    queryKey: [
      ...ADMIN_REQUESTS_KEY,
      filters.q.trim(),
      filters.status,
      filters.cityId,
      filters.tradeId,
      page,
    ],
    queryFn: () => api<Page<AdminRequestRow>>(`/admin/requests?${params}`),
    staleTime: 15_000,
    // Stepping through pages must not blink the list back to a skeleton.
    placeholderData: (previous) => previous,
  })
}

export function useAdminRequest(requestId: number | null) {
  return useQuery({
    queryKey: [...ADMIN_REQUESTS_KEY, 'one', requestId],
    queryFn: () => api<AdminRequest>(`/admin/requests/${requestId}`),
    enabled: requestId !== null,
  })
}

/** The one thing this screen can change, so the one thing that writes an
 *  audit row — which A8 reads. */
export function useCancelRequest() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ id, reason }: { id: number; reason: string }) =>
      api<AdminRequest>(`/admin/requests/${id}/cancel`, {
        method: 'POST',
        body: { reason },
      }),
    onSuccess: (request) => {
      queryClient.setQueryData([...ADMIN_REQUESTS_KEY, 'one', request.id], request)
      void queryClient.invalidateQueries({ queryKey: ADMIN_REQUESTS_KEY })
      void queryClient.invalidateQueries({ queryKey: AUDIT_KEY })
      void queryClient.invalidateQueries({ queryKey: STATS_KEY })
    },
  })
}

File: api/app/schemas/admin_requests.py
"""A4 — what the support browser renders."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.core.enums import (
    DisputeStatus,
    JobStatus,
    OfferStatus,
    RequestStatus,
    Urgency,
)
from app.schemas.catalog import CityOut, TradeOut
from app.schemas.common import ApiModel


class RequestPersonOut(ApiModel):
    """Somebody on a request, as support needs them.

    The phone is here and nowhere else in the admin API by accident: A4 exists
    because a person is *on* the phone, and an admin who has to leave the
    screen to find the number is an admin who reads it out wrong.
    """

    id: int
    full_name: str
    phone: str


class RequestRowOut(ApiModel):
    """One line in the list. Everything a support agent scans by."""

    id: int
    title: str
    status: RequestStatus
    urgency: Urgency
    offers_count: int
    created_at: datetime

    client: RequestPersonOut
    trade: TradeOut
    city: CityOut


class RequestOfferOut(ApiModel):
    id: int
    price_centimes: int
    message: str | None
    status: OfferStatus
    created_at: datetime
    #: Frozen when the offer was accepted, so a later price change never
    #: rewrites what this lead actually cost. Null on an offer nobody took.
    lead_fee_centimes: int | None
    provider: RequestPersonOut
    provider_id: int


class RequestJobOut(ApiModel):
    id: int
    status: JobStatus
    agreed_price_centimes: int
    started_at: datetime | None
    finished_at: datetime | None
    confirmed_at: datetime | None
    cancelled_at: datetime | None
    cancelled_by: str | None
    cancel_reason: str | None
    provider: RequestPersonOut | None


class RequestDisputeOut(ApiModel):
    id: int
    status: DisputeStatus
    reason: str
    opened_by_id: int


class RequestDetailOut(ApiModel):
    """The whole story of one request, in one response."""

    id: int
    title: str
    description: str
    status: RequestStatus
    urgency: Urgency
    address: str
    budget_min_centimes: int | None
    budget_max_centimes: int | None
    photos: int
    offers_count: int

    created_at: datetime
    expires_at: datetime | None
    cancelled_at: datetime | None
    cancel_reason: str | None

    client: RequestPersonOut
    trade: TradeOut
    city: CityOut

    offers: list[RequestOfferOut]
    job: RequestJobOut | None
    dispute: RequestDisputeOut | None

    #: Whether A4's one action is available. Answered here rather than derived
    #: on the screen from `status`, so the button and the API agree about what
    #: "open" means.
    can_cancel: bool


class CancelRequestIn(BaseModel):
    #: Mandatory. Somebody reads this in three months and needs to know why a
    #: client's request vanished; "cancelled by an admin" is not a why.
    reason: str = Field(min_length=1, max_length=500)
